File: wpimath/src/test/java/org/wpilib/math/filter/BiquadFilterTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.math.filter;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Random;
import org.junit.jupiter.api.Test;

class BiquadFilterTest {
  @Test
  void passThroughTest() {
    BiquadFilter filter = new BiquadFilter(new BiquadFilter.Section(1.0, 0.0, 0.0, 0.0, 0.0));
    Random rng = new Random(42);
    for (int i = 0; i < 200; i++) {
      double x = (rng.nextDouble() - 0.5) * 200.0;
      assertEquals(x, filter.calculate(x), 0.0);
    }
  }

  @Test
  void firstOrderMatchesSinglePoleIIRTest() {
    // SinglePoleIIR: y[n] = (1-g) x[n] + g y[n-1], g = exp(-dt/T)
    // As biquad: {1-g, 0, 0, -g, 0}
    final double timeConstant = 0.015915;
    final double period = 0.005;
    final double g = Math.exp(-period / timeConstant);

    LinearFilter linear = LinearFilter.singlePoleIIR(timeConstant, period);
    BiquadFilter biquad = new BiquadFilter(new BiquadFilter.Section(1.0 - g, 0.0, 0.0, -g, 0.0));

    Random rng = new Random(7);
    for (int i = 0; i < 1000; i++) {
      double x = (rng.nextDouble() - 0.5) * 100.0;
      double yLin = linear.calculate(x);
      double yBiq = biquad.calculate(x);
      assertEquals(yLin, yBiq, 1e-12, "sample " + i);
    }
  }

  @Test
  void butterworth4thOrderLowPassTest() {
    // scipy.signal.butter(4, 50.0, btype='low', fs=1000.0, output='sos')
    BiquadFilter filter =
        new BiquadFilter(
            new BiquadFilter.Section(
                0.00041659920440659937,
                0.0008331984088131987,
                0.00041659920440659937,
                -1.4796742169311934,
                0.5558215432824889),
            new BiquadFilter.Section(1.0, 2.0, 1.0, -1.7009643319435257, 0.7884997398152979));

    double[] expected = {
      0.00041659920440659937,
      0.0029914483065925663,
      0.010405740533503665,
      0.024092655231875183,
      0.04300386328531425,
      0.06442081415630327,
      0.08518000836484753,
      0.10245740377665029,
      0.1142030744642985,
      0.11931076610150239,
      0.11759868177262096,
      0.10966797135549569,
      0.09669445862739445,
      0.08019770689053446,
      0.06182021082200691,
      0.04313888252371942,
      0.025521549440937964,
      0.01003324447867498,
      -0.002609160074363505,
      -0.01203989315971688,
      -0.018213508615082776,
      -0.021350392922805498,
      -0.02186860712506684,
      -0.020311212598085788,
      -0.01727668431352673,
      -0.013358111475758645,
      -0.009094876704322833,
      -0.004938595712161598,
      -0.0012334395430879353,
      0.0017903545884787877,
    };

    for (int i = 0; i < expected.length; i++) {
      double x = (i == 0) ? 1.0 : 0.0;
      double y = filter.calculate(x);
      assertEquals(expected[i], y, 1e-10, "sample " + i);
    }
  }

  @Test
  void notch60HzTest() {
    // scipy.signal.iirnotch(60, Q=10, fs=1000) via tf2sos
    BiquadFilter filter =
        new BiquadFilter(
            new BiquadFilter.Section(
                0.9814970254751076,
                -1.8251457105120343,
                0.9814970254751076,
                -1.8251457105120341,
                0.9629940509502151));

    final double fs = 1000.0;
    final int samples = 1000;
    double[] output = new double[samples];
    double[] input = new double[samples];
    for (int n = 0; n < samples; n++) {
      double t = n / fs;
      input[n] = Math.sin(2.0 * Math.PI * 10.0 * t) + Math.sin(2.0 * Math.PI * 60.0 * t);
      output[n] = filter.calculate(input[n]);
    }

    assertEquals(-0.017355123579818322, output[500], 1e-10);
    assertEquals(-0.08007594066581347, output[999], 1e-10);

    // Basic DFT at 10 Hz and 60 Hz over steady-state window.
    final int window = 512;
    double atten60 = attenuationDb(output, input, 60.0, fs, samples - window, window);
    double atten10 = attenuationDb(output, input, 10.0, fs, samples - window, window);

    assertTrue(atten60 < -40.0, "60 Hz attenuation too weak: " + atten60);
    assertTrue(atten10 > -0.5, "10 Hz passband loss too large: " + atten10);
  }

  @Test
  void order8ButterworthMatchesScipyTest() {
    // scipy.signal.butter(8, 100.0, btype='low', fs=1000.0, output='sos')
    BiquadFilter filter =
        new BiquadFilter(
            new BiquadFilter.Section(
                2.3959644103776166e-05,
                4.791928820755233e-05,
                2.3959644103776166e-05,
                -1.0263514742610553,
                0.26864019099379005),
            new BiquadFilter.Section(1.0, 2.0, 1.0, -1.0868584613628944, 0.343430940165366),
            new BiquadFilter.Section(1.0, 2.0, 1.0, -1.2197253651240232, 0.5076634651740437),
            new BiquadFilter.Section(1.0, 2.0, 1.0, -1.4515795942478362, 0.794251053241888));

    final int N = 500;
    final double fs = 1000.0;
    final double f0 = 1.0;
    final double f1 = 200.0;
    final double t1 = (N - 1) / fs;
    final double k = (f1 - f0) / t1;

    int[] spotIdx = {10, 50, 100, 250, 499};
    double[] expected = {
      0.8950675041062186,
      -0.7902247252134351,
      0.1716891991372734,
      0.05240058121316523,
      -0.0016952227415119995,
    };
    double[] got = new double[spotIdx.length];

    int j = 0;
    for (int n = 0; n < N; n++) {
      double t = n / fs;
      double phase = 2.0 * Math.PI * (f0 * t + 0.5 * k * t * t);
      double x = Math.cos(phase);
      double y = filter.calculate(x);
      if (j < spotIdx.length && n == spotIdx[j]) {
        got[j++] = y;
      }
    }

    for (int i = 0; i < expected.length; i++) {
      assertEquals(expected[i], got[i], 1e-10, "sample index " + spotIdx[i]);
    }
  }

  @Test
  void resetZerosStateTest() {
    BiquadFilter filter =
        new BiquadFilter(
            new BiquadFilter.Section(
                0.00041659920440659937,
                0.0008331984088131987,
                0.00041659920440659937,
                -1.4796742169311934,
                0.5558215432824889),
            new BiquadFilter.Section(1.0, 2.0, 1.0, -1.7009643319435257, 0.7884997398152979));

    for (int i = 0; i < 50; i++) {
      filter.calculate(1.0);
    }
    assertNotEquals(0.0, filter.lastValue());

    filter.reset();
    assertEquals(0.0, filter.lastValue(), 0.0);

    double y = filter.calculate(1.0);
    assertEquals(0.00041659920440659937, y, 1e-12);
  }

  @Test
  void resetToSteadyStateTest() {
    BiquadFilter filter =
        new BiquadFilter(
            new BiquadFilter.Section(
                0.00041659920440659937,
                0.0008331984088131987,
                0.00041659920440659937,
                -1.4796742169311934,
                0.5558215432824889),
            new BiquadFilter.Section(1.0, 2.0, 1.0, -1.7009643319435257, 0.7884997398152979));

    final double input = 3.0;
    filter.reset(input);

    assertEquals(input, filter.lastValue(), 1e-12);
    assertEquals(input, filter.calculate(input), 1e-12);
    for (int i = 0; i < 20; i++) {
      assertEquals(input, filter.calculate(input), 1e-12);
    }
  }

  @Test
  void dcGainConvergesTest() {
    BiquadFilter filter =
        new BiquadFilter(
            new BiquadFilter.Section(
                0.00041659920440659937,
                0.0008331984088131987,
                0.00041659920440659937,
                -1.4796742169311934,
                0.5558215432824889),
            new BiquadFilter.Section(1.0, 2.0, 1.0, -1.7009643319435257, 0.7884997398152979));

    final double input = 2.5;
    double y = 0.0;
    for (int i = 0; i < 500; i++) {
      y = filter.calculate(input);
    }
    assertEquals(input, y, 1e-6);
  }

  @Test
  void numSectionsTest() {
    BiquadFilter one = new BiquadFilter(new BiquadFilter.Section(1.0, 0.0, 0.0, 0.0, 0.0));
    assertEquals(1, one.numSections());

    BiquadFilter three =
        new BiquadFilter(
            new BiquadFilter.Section(1.0, 0.0, 0.0, 0.0, 0.0),
            new BiquadFilter.Section(1.0, 0.0, 0.0, 0.0, 0.0),
            new BiquadFilter.Section(1.0, 0.0, 0.0, 0.0, 0.0));
    assertEquals(3, three.numSections());
  }

  @Test
  void emptyCascadeThrowsTest() {
    assertThrows(IllegalArgumentException.class, () -> new BiquadFilter());
  }

  private static double attenuationDb(
      double[] out, double[] in, double freq, double fs, int start, int window) {
    double magOut = dftMag(out, freq, fs, start, window);
    double magIn = dftMag(in, freq, fs, start, window);
    return 20.0 * Math.log10(magOut / magIn);
  }

  private static double dftMag(double[] sig, double freq, double fs, int start, int window) {
    double re = 0.0;
    double im = 0.0;
    for (int n = 0; n < window; n++) {
      double phase = 2.0 * Math.PI * freq * n / fs;
      re += sig[start + n] * Math.cos(phase);
      im -= sig[start + n] * Math.sin(phase);
    }
    return Math.hypot(re, im);
  }
}

File: wpimath/src/test/native/cpp/filter/BiquadFilterTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/math/filter/BiquadFilter.hpp"

#include <array>
#include <cmath>
#include <cstddef>
#include <numbers>
#include <random>
#include <vector>

#include <gtest/gtest.h>

#include "wpi/math/filter/LinearFilter.hpp"
#include "wpi/units/time.hpp"

using wpi::math::BiquadFilter;

TEST(BiquadFilterTest, PassThrough) {
  BiquadFilter filter({{1.0, 0.0, 0.0, 0.0, 0.0}});

  std::mt19937 rng(42);
  std::uniform_real_distribution<double> dist(-100.0, 100.0);
  for (int i = 0; i < 200; ++i) {
    double x = dist(rng);
    EXPECT_DOUBLE_EQ(filter.Calculate(x), x);
  }
}

TEST(BiquadFilterTest, FirstOrderMatchesSinglePoleIIR) {
  // SinglePoleIIR: y[n] = (1-g) x[n] + g y[n-1], g = exp(-dt/T)
  // As biquad:     {1-g, 0, 0, -g, 0}
  constexpr double kTimeConstant = 0.015915;
  constexpr double kPeriod = 0.005;
  double g = std::exp(-kPeriod / kTimeConstant);

  auto linear =
      wpi::math::LinearFilter<double>::SinglePoleIIR(kTimeConstant, 5_ms);
  BiquadFilter biquad({{1.0 - g, 0.0, 0.0, -g, 0.0}});

  std::mt19937 rng(7);
  std::uniform_real_distribution<double> dist(-50.0, 50.0);
  for (int i = 0; i < 1000; ++i) {
    double x = dist(rng);
    double y_lin = linear.Calculate(x);
    double y_biq = biquad.Calculate(x);
    EXPECT_NEAR(y_lin, y_biq, 1e-12);
  }
}

TEST(BiquadFilterTest, Butterworth4thOrderLowPass) {
  // scipy.signal.butter(4, 50.0, btype='low', fs=1000.0, output='sos')
  BiquadFilter filter({
      {0.00041659920440659937, 0.0008331984088131987, 0.00041659920440659937,
       -1.4796742169311934, 0.5558215432824889},
      {1.0, 2.0, 1.0, -1.7009643319435257, 0.7884997398152979},
  });

  // Impulse response, first 30 samples, from scipy.signal.sosfilt.
  constexpr std::array<double, 30> kExpected = {
      0.00041659920440659937, 0.0029914483065925663,  0.010405740533503665,
      0.024092655231875183,   0.04300386328531425,    0.06442081415630327,
      0.08518000836484753,    0.10245740377665029,    0.1142030744642985,
      0.11931076610150239,    0.11759868177262096,    0.10966797135549569,
      0.09669445862739445,    0.08019770689053446,    0.06182021082200691,
      0.04313888252371942,    0.025521549440937964,   0.01003324447867498,
      -0.002609160074363505,  -0.01203989315971688,   -0.018213508615082776,
      -0.021350392922805498,  -0.02186860712506684,   -0.020311212598085788,
      -0.01727668431352673,   -0.013358111475758645,  -0.009094876704322833,
      -0.004938595712161598,  -0.0012334395430879353, 0.0017903545884787877,
  };

  for (size_t i = 0; i < kExpected.size(); ++i) {
    double x = (i == 0) ? 1.0 : 0.0;
    double y = filter.Calculate(x);
    EXPECT_NEAR(y, kExpected[i], 1e-10) << "sample " << i;
  }
}

TEST(BiquadFilterTest, Notch60Hz) {
  // scipy.signal.iirnotch(60.0, Q=10.0, fs=1000.0), converted via tf2sos
  BiquadFilter filter({
      {0.9814970254751076, -1.8251457105120343, 0.9814970254751076,
       -1.8251457105120341, 0.9629940509502151},
  });

  constexpr double kFs = 1000.0;
  constexpr int kSamples = 1000;
  std::vector<double> output(kSamples);
  for (int n = 0; n < kSamples; ++n) {
    double t = n / kFs;
    double x = std::sin(2.0 * std::numbers::pi * 10.0 * t) +
               std::sin(2.0 * std::numbers::pi * 60.0 * t);
    output[n] = filter.Calculate(x);
  }

  // Spot-check against scipy.signal.sosfilt outputs
  EXPECT_NEAR(output[500], -0.017355123579818322, 1e-10);
  EXPECT_NEAR(output[999], -0.08007594066581347, 1e-10);

  // Attenuation check via a basic DFT at 10 Hz and 60 Hz over the last 512
  // samples (in steady state). 60 Hz should be strongly attenuated, 10 Hz
  // should pass almost untouched.
  constexpr int kWindow = 512;
  auto bin = [&](const std::vector<double>& sig, double freq) {
    double re = 0.0;
    double im = 0.0;
    for (int n = 0; n < kWindow; ++n) {
      double x = sig[kSamples - kWindow + n];
      double phase = 2.0 * std::numbers::pi * freq * n / kFs;
      re += x * std::cos(phase);
      im -= x * std::sin(phase);
    }
    return std::hypot(re, im);
  };

  std::vector<double> input(kSamples);
  for (int n = 0; n < kSamples; ++n) {
    double t = n / kFs;
    input[n] = std::sin(2.0 * std::numbers::pi * 10.0 * t) +
               std::sin(2.0 * std::numbers::pi * 60.0 * t);
  }

  double in10 = bin(input, 10.0);
  double in60 = bin(input, 60.0);
  double out10 = bin(output, 10.0);
  double out60 = bin(output, 60.0);

  double atten60_dB = 20.0 * std::log10(out60 / in60);
  double atten10_dB = 20.0 * std::log10(out10 / in10);

  EXPECT_LT(atten60_dB, -40.0) << "60 Hz not sufficiently attenuated";
  EXPECT_GT(atten10_dB, -0.5) << "10 Hz passband loss too large";
}

TEST(BiquadFilterTest, Order8ButterworthMatchesScipy) {
  // High-order filter = 4 biquads. This test exists to prove that the SOS
  // (Direct Form II Transposed) implementation is numerically correct at the
  // orders that a flattened-polynomial LinearFilter cannot reliably run.
  //
  // scipy.signal.butter(8, 100.0, btype='low', fs=1000.0, output='sos')
  BiquadFilter filter({
      {2.3959644103776166e-05, 4.791928820755233e-05, 2.3959644103776166e-05,
       -1.0263514742610553, 0.26864019099379005},
      {1.0, 2.0, 1.0, -1.0868584613628944, 0.343430940165366},
      {1.0, 2.0, 1.0, -1.2197253651240232, 0.5076634651740437},
      {1.0, 2.0, 1.0, -1.4515795942478362, 0.794251053241888},
  });

  // Linear chirp from 1 Hz to 200 Hz over 500 samples at 1 kHz.
  // Matches scipy.signal.chirp(t, f0=1, f1=200, t1=t[-1], method='linear').
  constexpr int kSamples = 500;
  constexpr double kFs = 1000.0;
  constexpr double kF0 = 1.0;
  constexpr double kF1 = 200.0;
  const double t1 = (kSamples - 1) / kFs;
  const double k = (kF1 - kF0) / t1;

  std::array<double, 5> spot_samples{};
  constexpr std::array<int, 5> kSpotIndices{10, 50, 100, 250, 499};
  constexpr std::array<double, 5> kExpected{
      0.8950675041062186,  -0.7902247252134351,    0.1716891991372734,
      0.05240058121316523, -0.0016952227415119995,
  };

  size_t spot_idx = 0;
  for (int n = 0; n < kSamples; ++n) {
    double t = n / kFs;
    double phase = 2.0 * std::numbers::pi * (kF0 * t + 0.5 * k * t * t);
    double x = std::cos(phase);
    double y = filter.Calculate(x);

    if (spot_idx < kSpotIndices.size() && n == kSpotIndices[spot_idx]) {
      spot_samples[spot_idx++] = y;
    }
  }

  for (size_t i = 0; i < kExpected.size(); ++i) {
    EXPECT_NEAR(spot_samples[i], kExpected[i], 1e-10)
        << "sample index " << kSpotIndices[i];
  }
}

TEST(BiquadFilterTest, ResetZerosState) {
  BiquadFilter filter({
      {0.00041659920440659937, 0.0008331984088131987, 0.00041659920440659937,
       -1.4796742169311934, 0.5558215432824889},
      {1.0, 2.0, 1.0, -1.7009643319435257, 0.7884997398152979},
  });

  for (int i = 0; i < 50; ++i) {
    filter.Calculate(1.0);
  }
  EXPECT_NE(filter.LastValue(), 0.0);

  filter.Reset();
  EXPECT_DOUBLE_EQ(filter.LastValue(), 0.0);

  // First call after Reset should behave like the filter starts fresh —
  // matches the impulse-response first sample.
  double y = filter.Calculate(1.0);
  EXPECT_NEAR(y, 0.00041659920440659937, 1e-12);
}

TEST(BiquadFilterTest, ResetToSteadyState) {
  // DC gain of each section is (b0+b1+b2)/(1+a1+a2). After Reset(value),
  // Calculate(value) should immediately return value * cascade_DC_gain.
  BiquadFilter filter({
      {0.00041659920440659937, 0.0008331984088131987, 0.00041659920440659937,
       -1.4796742169311934, 0.5558215432824889},
      {1.0, 2.0, 1.0, -1.7009643319435257, 0.7884997398152979},
  });

  constexpr double kInput = 3.0;
  filter.Reset(kInput);

  // Cascade DC gain for a Butterworth LP is 1.0, so output should equal input.
  EXPECT_NEAR(filter.LastValue(), kInput, 1e-12);
  double y = filter.Calculate(kInput);
  EXPECT_NEAR(y, kInput, 1e-12);

  // And remain at steady state
  for (int i = 0; i < 20; ++i) {
    EXPECT_NEAR(filter.Calculate(kInput), kInput, 1e-12);
  }
}

TEST(BiquadFilterTest, DCGainConverges) {
  BiquadFilter filter({
      {0.00041659920440659937, 0.0008331984088131987, 0.00041659920440659937,
       -1.4796742169311934, 0.5558215432824889},
      {1.0, 2.0, 1.0, -1.7009643319435257, 0.7884997398152979},
  });

  constexpr double kInput = 2.5;
  double y = 0.0;
  for (int i = 0; i < 500; ++i) {
    y = filter.Calculate(kInput);
  }
  EXPECT_NEAR(y, kInput, 1e-6);  // Butterworth LP has DC gain 1
}

TEST(BiquadFilterTest, NumSections) {
  BiquadFilter one({{1.0, 0.0, 0.0, 0.0, 0.0}});
  EXPECT_EQ(one.NumSections(), 1u);

  BiquadFilter three({
      {1.0, 0.0, 0.0, 0.0, 0.0},
      {1.0, 0.0, 0.0, 0.0, 0.0},
      {1.0, 0.0, 0.0, 0.0, 0.0},
  });
  EXPECT_EQ(three.NumSections(), 3u);
}

TEST(BiquadFilterTest, EmptyCascadeThrows) {
  EXPECT_THROW(
      {
        std::vector<BiquadFilter::Section> sections;
        std::span<const BiquadFilter::Section> empty{sections};
        BiquadFilter filter{empty};
      },
      std::runtime_error);
}
